### PR TITLE
Zhijxu/cast propagation softmax

### PR DESCRIPTION
**Pull request by zhijxu-MS**
enhance cast-propagation for "softmax can be put at fp16 when data flow is cast-to-fp32 > softmax > cast-to-fp16"

this optimization can save gpu memory and have performance gain

### CODE REVIEW

File: onnxruntime/core/optimizer/propagate_cast_ops.cc [hunk 1]
```diff
@@ -127,8 +127,43 @@ static bool IsRelevantOutput(const Node* node, const NodeArg* output) {
...
+  const Node* output_node = &(*node.OutputNodesBegin());
+  if (!(output_node && IsCastTo(output_node, TensorProto::FLOAT16)))
+    return false;
+
+  return true;
+}
+
 // Check whether the given opcode is fp16 allowed for the given level of optimization.
-static bool IsFP16Allow(const std::string& op_type, size_t level, const FP16AllowOps& fp16_allow_level0_ops) {
+static bool IsFP16Allow(const std::string& op_type, size_t level, const FP16AllowOps& fp16_allow_level0_ops, const Node* node = nullptr) {
```
Comment (pengwa) [inline onnxruntime/core/optimizer/propagate_cast_ops.cc:166]: why not make `const Node* node` be a required param, and remove `const std::string& op_type`?

File: onnxruntime/test/optimizer/graph_transform_test.cc [hunk 2]
```diff
@@ -5157,6 +5157,74 @@ TEST_F(GraphTransformationTests, PropagateCastOpsTests_Gelu) {
 
 #endif
 
+#ifdef ENABLE_TRAINING_CORE
```
Comment (pengwa) [inline onnxruntime/test/optimizer/graph_transform_test.cc:5160]: you can merge this macro with pervious test case: 

"#endif

#ifdef ENABLE_TRAINING_CORE"

File: onnxruntime/test/optimizer/graph_transform_test.cc [hunk 3]
```diff
@@ -5157,6 +5157,74 @@ TEST_F(GraphTransformationTests, PropagateCastOpsTests_Gelu) {
...
+TEST_F(GraphTransformationTests, PropagateCastOpsTests_Softmax) {
+  using Strategy = GraphTransformerConfiguration::PropagateCastOpsConfiguration::Strategy;
+  {
+    auto build_test_case = [&](ModelTestBuilder& builder) {
```
Comment (pengwa) [inline onnxruntime/test/optimizer/graph_transform_test.cc:5164]: nit: 
```suggestion
    auto build_test_case = [](ModelTestBuilder& builder) {
```

**Review by pengwa — APPROVED** [review]

**Review by mindest — APPROVED** [review]

**Review by baijumeswani — APPROVED** [review]

**Comment by tianleiwu** [comment]
Does this PR have assumption that the execution provider implemented both FP32 and PF16 Softmax. What if a provider only implement FP32 Softmax?

**Comment by zhijxu-MS** [comment]
> Does this PR have assumption that the execution provider implemented both FP32 and PF16 Softmax. What if a provider only implement FP32 Softmax?

this optimization only includes in training scenario, so softmax should have fp16 cure implementation; also this optimization already does the assumption, such as the [InsertAndReduce](https://github.com/microsoft/onnxruntime/blob/9407c3270c10352ad7e55a9bcb16f733afb23c79/onnxruntime/core/optimizer/propagate_cast_ops.cc#L26) stategy will try to put all op into fp16

File: onnxruntime/core/optimizer/propagate_cast_ops.cc [hunk 4]
```diff
@@ -127,8 +127,45 @@ static bool IsRelevantOutput(const Node* node, const NodeArg* output) {
...
+    // Allow cast nodes with same input and output type float/float16 to eliminate such casts.
+    return (input_data_type == TensorProto::FLOAT16 || input_data_type == TensorProto::FLOAT) &&
+           attr_hit->second.i() == static_cast<int64_t>(data_type);
+  }
+  return false;
+}
+
+// when node is softmax, and its input comes from a cast-to-fp32 node, and its output is only consumed by one cast-to-fp16 node, then we can treat it as a fp16-allowed op,
+// as ort's softmax implementation already does the necessary cast logic, for example do reduce sum at fp32
+static bool SoftmaxCanBeFP16(const Node& node) {
```
[the remark this answers is not in the document]
Comment (zhijxu-MS) [reply onnxruntime/core/optimizer/propagate_cast_ops.cc:148]: good point! we could do it this way when we figure out more ops can work with the idea.
